api/src/main/java/javax/config/Config.java

```diff
@@ -127,4 +129,12 @@
      * @return all currently registered {@link ConfigSource configsources} sorted with descending ordinal and ConfigSource name
      */
     Iterable<ConfigSource> getConfigSources();
+
+    /**
+     * A user can register a lambda which gets notified when any configured value
+     * got changed. The parameter are the config key names which did change.
+     *
+     * @param configChangedListener
+     */
+    void registerConfigChangedListener(Consumer<Set<String>> configChangedListener);
```

Comment:
The name is not consistent with the `setOnAttributeChange` below.

Maybe `registerOnConfigChange(Consumer<Set<String>> reportConfigChange)`?

`setX` implies that there is only 1 that will be kept, `registerX` implies that there can be many.

Reply:
I would rather call it `registerConfigChangeListener` than `registerConfigChangedListener`. It is more consistent with elements of the JDK like `ChangeListener`.

Reply:
Would it make sense to also (maybe optionally) specify which keys you want to be notified of? 

I think typically there will be a small subset of keys that you actually want to listen for changes to. This could also be then passed down to config sources so they can watch changes for only the keys you are interested in.

Reply:
we thought about that. In @tomas-langer PR https://github.com/eclipse/ConfigJSR/pull/46 this is split up in 3 methods.

Personally I think it will depend on how we plan to enhance config in the future.
If we will also add variableReplacement and fallbackChains then the value of my.key.url might e.g. be depending on another key. And the user code sometimes doesn't even know which other keys do affect him.

Consider the following use case:

my.key.url = http://${key.server.name}/some/path
key.server.name=myserver.net

So config.getValue("my.key.url") will return something different if key.server.name changes, right?
But if you only listen to changes on my.key.url you will NEVER get notified.
Same happens with fallbackChains and other advanced use cases.
Got me?

Reply:
Yes that's true and this should definitely support that. In any case, however, the config implementation will probably need to know the dependencies between different keys so it can accurately figure out which keys changed, whether you watch for every key or only a few.

PR #46 would be a good way to take this route.